### DISCUSSION TITLE
[General] Is it possible to run Era VM in isolated env?

### DISCUSSION THREAD

**Discussion by alexhooketh** [zkSync-Community-Hub/zksync-developers#213] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I want to make a script that requires running Era VM in isolated environment to execute necessary computations locally, I look for functionality like py-evm or revm. I prefer using rust but can use any language. I didn't find anything related to this in docs, so I'm not sure if it's possible. If it's not, is it possible to implement such sandbox? If there are specs of VM or something, I'd like to try to make it myself

### What is the specific issue or error you're encountering?

I didn't find anything related

### Can you share the error messages or logs you're receiving, if any?

-

### Have you made any recent changes to the contract before encountering this issue?

-

### Are there any external libraries or contracts that your contract interacts with?

-

### Can you provide the relevant portions of your contract code where the issue is occurring?

-

### Have you tried to isolate the problem, and if so, what were the results?

-

### What steps have you already taken to try to resolve the issue?

-

**Comment by dutterbutter** [comment, marked as answer]
Hey 👋 

Not exactly sure what the outcome you are looking for BUT we are currently working on [era-revm](https://github.com/matter-labs/era-revm/tree/main) which acts as a translation layer for our ongoing foundry support. This but no means is production-ready, lacks testing and has a lot of assumptions in it currently but is likely the closest thing to what you are looking for. We make use of it by calling it from our foundry fork [here](https://github.com/matter-labs/foundry-zksync/blob/c514051b9dfb7babeb5b896962db02381f1e1382/crates/evm/src/executor/backend/mod.rs#L774). Not exactly sure if that is helpful for your purpose?

**Reply by alexhooketh** [reply]
Yes, that's pretty much what I looked for. Sad that it's not ready yet, I'll try to look through the code and maybe contribute. Thanks!

**Comment by rietracerdanchri** [comment]
To execute computations locally in an isolated environment similar to py-evm or revm, you can explore building a custom virtual machine (VM) using Rust or another programming language. You would need to understand the specifications of the VM, such as the EVM (Ethereum Virtual Machine), and implement its functionality, including opcodes, gas handling, stack operations, and memory management. While there may not be existing libraries specifically tailored for this purpose in the zkSync ecosystem, you can refer to the Ethereum Yellow Paper for detailed specifications of the EVM and leverage Rust's capabilities to create your custom VM. This approach allows you to execute smart contract computations locally in a sandboxed environment for testing and development purposes.

**Comment by marqbassmier** [comment]
To execute computations locally in an isolated environment similar to Py-EVM or REVM, you can explore implementing a custom virtual machine (VM) using Rust or another language. While there may not be specific documentation for running Era VM in isolation, you can refer to the existing VM specifications and protocols to develop your own sandbox environment for executing computations. Consider leveraging existing VM specifications and protocols as a reference to implement your solution.

**Comment by meutrandeesni** [comment]
It's possible to execute computations locally using zkSync's zkEVM. While there isn't an official SDK or library for it yet, you can refer to zkSync's documentation and the EVM specifications to implement your own sandbox in Rust or any other language. This approach would allow you to run Era VM in an isolated environment for your script.

**Comment by porbotufibr** [comment]
To execute computations in an isolated environment similar to Py-EVM or ReVM, consider implementing a sandboxed Era VM using Rust or another language. While there may not be existing documentation or libraries, you can explore Era VM specifications and build your own sandboxed environment for local execution.

**Comment by swelexdewi** [comment]
The user is seeking to implement a script that requires running the Era VM in an isolated environment for local computations. They are looking for functionality similar to py-evm or revm and are open to using any language, but prefer Rust. They have not encountered any specific issues or errors, but have not found any relevant information in the documentation. They are also interested in knowing if it's possible to implement such a sandbox and are willing to try creating it themselves if there are specs available. The user has not provided any code or repository links.

**Comment by sanddethe** [comment]
The user is looking to create a script that requires running Era VM in an isolated environment to execute necessary computations locally. They are looking for functionality like py-evm or revm and prefer to use Rust, but can use any language. They did not find anything related to this in the documentation and are unsure if it's possible. They are asking if it's possible to implement such a sandbox and if there are VM specs available.
